segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/SegmentAttributeIterator.java

```diff
@@ -0,0 +1,158 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.server.containers;
+
+import com.google.common.base.Preconditions;
+import io.pravega.segmentstore.contracts.Attributes;
+import io.pravega.segmentstore.server.AttributeIterator;
+import io.pravega.segmentstore.server.SegmentMetadata;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import lombok.NonNull;
+import lombok.val;
+
+/**
+ * Implements a {@link AttributeIterator} for a Segment, using both the Segment Attribute Index and Segment Metadata
+ * as sources.
+ *
+ * While the Segment Attribute Index is the final storage for all the Segment's Extended Attributes, the Segment Metadata
+ * also contains the Core Attributes and those Extended Attributes that have recently been changed. As such, if an
+ * Attribute exists in the Segment Metadata, it's value is the most recent and must be used (eventually that value will
+ * trickle down to the Attribute Index).
+ *
+ * Instances of this class iterate over the Attribute Index (within the specified bounds), and also include Attributes
+ * from the Segment Metadata where appropriate.
+ */
+class SegmentAttributeIterator implements AttributeIterator {
+    //region Members
+
+    private final AttributeIterator indexIterator;
+    private final ArrayDeque<Map.Entry<UUID, Long>> metadataAttributes;
```

Comment:
ArrayDequeue is not thread safe.

Reply:
Fixed this whole class to be thread safe.